_posts/2021-04-05-maven-bom.md

```diff
@@ -0,0 +1,316 @@
+      \- (org.springframework:spring-core:jar:5.2.13.RELEASE:compile - omitted for conflict with 5.3.5)
+```
+
+We started from two dependencies, and in this output, we find out that Maven pulled additional dependencies. These additional dependencies are simply transitive.
+
+We can see that there are different versions of the same dependency in the tree. For example, there are two versions of the `spring-beans` dependency:`5.2.13.RELEASE` and `5.3.5`.
+
+Maven has resolved this version conflict, but how? What does *omitted for duplicate* and *omitted for conflict* mean?
+
+### How Does Maven Resolve Version Conflicts?
+
+The first thing to know is that Maven can't sort versions: The versions are arbitrary strings and may not follow a strict semantic sequence. For example, if we have two versions `1.2` and `1.11`, we know that `1.11` comes after `1.2` but the String comparison gives `1.11` before `1.2`. Other version values can be `1.1-rc1` or `1.1-FINAL`, that's why sorting versions by Maven is not a solution.  
+
+**That means that Maven doesn't know which version is newer or older and cannot choose to always take the newest version.**
+
+Second, **Maven takes the approach of the nearest transitive dependency in the tree depth and the first in resolution**. To understand this, let's look at an example:
+
+We start with a POM file having some dependencies with transitive dependencies (to make it short, all the dependencies will be represented by the letter D):
+
+- `D1(v1) -> D11(v11) -> D12(v12) -> DT(v1.3)`
+- `D2(v2) -> DT(v1.2)`
+- `D3(v3) -> D31(v31) -> DT(v1.0)`
+- `D4(v4) -> DT(v1.5)`
+
+Note that each of the direct dependencies pulls in a different version of the `DT` dependency.
+
+Maven will create a dependency tree and following the criteria mentioned above, a dependency will be selected for `DT`:
+
+![Transitive dependency resolution](/assets/img/posts/bom/transitive-dependencies-resolution.png)
+
+We note that the resolution order played a major role in choosing the `DT` dependency since the `v1.2` and `v1.5` had the same depth, but `v1.2` came first in the resolution order. So even if `v1.2` is not the last version of `DT`, Maven chose it to work with.
+
+If we wanted to use version `v1.5` in this case, we could simply add the dependency `D4` before `D2` in our POM file. In this case, `v1.5` will be first in terms of resolution order and Maven will select it.
+
+So, to help us understand the dependency tree result from above, Maven indicates for each transitive dependency why it was omitted:
+
+- **"omitted for duplicate"** means that the same dependency with the *same* version is declared at another place in the dependency tree higher in resolution order or depth and Maven prefers that one.
+- **"omitted for conflict"** means that the same dependency with a *different* version is declared at another place in the dependency tree higher in resolution order or depth and Maven prefers that one.
+
```

Comment:
For these two cases it's the opposite: the other dependency chosen by Maven is lower in resolution order and nearer in depth

Reply:
@bakic hmm, I understand if Maven says "omitted", it means that THIS dependency (i.e. the dependency annotated with "omitted")  is lower in resolution order and the OTHER dependency is higher in resolution order. Otherwise Maven would choose the OTHER dependency, wouldn't it?

Reply:
@thombergs it's the opposite, the OTHER dependency is the one chosen by Maven which is LOWER in resolution order and the NEAREST in depth. The one omitted is the opposite of it.
From the article: Maven takes the approach of the nearest transitive dependency in the tree depth and the first in resolution

Reply:
@thombergs ah I see, I think that we have different interpretations of the lower and higher terms, but our comprehension of the mechanism of Maven is the same. By lower you mean the first in resolution order and higher the one that comes after, isn't it? These terms may be confusing, don't you think that? in fact, we are having this discussion because of them 😅